v2/tools/generator/internal/codegen/pipeline/add_secrets.go

```diff
@@ -0,0 +1,134 @@
+func AddSecrets(_ *config.Configuration) Stage { // TODO: Will need this configuration parameter eventually to allow config overrides
+	return MakeStage(
+		AddSecretsStageID,
+		"Replace properties flagged as secret with genruntime.SecretReference",
+		func(ctx context.Context, state *State) (*State, error) {
+
+			updatedSpecs, err := transformSpecSecrets(state.Types())
+			if err != nil {
+				return nil, err
+			}
+
+			updatedStatuses, err := removeStatusSecrets(state.Types())
+			if err != nil {
+				return nil, err
+			}
+
+			result := state.Types().OverlayWith(astmodel.TypesDisjointUnion(updatedSpecs, updatedStatuses))
+			return state.WithTypes(result), nil
+		})
+}
+
+func transformSpecSecrets(types astmodel.Types) (astmodel.Types, error) {
+	specVisitor := astmodel.TypeVisitorBuilder{
+		VisitObjectType: transformSecretProperties,
+	}.Build()
+
+	specTypes, err := astmodel.FindSpecConnectedTypes(types)
+	if err != nil {
+		return nil, errors.Wrap(err, "couldn't find all spec types")
+	}
+
+	result := make(astmodel.Types)
+
+	for _, def := range specTypes {
+		updatedDef, err := specVisitor.VisitDefinition(def, nil)
+		if err != nil {
+			return nil, errors.Wrapf(err, "visiting type %q", def.Name())
+		}
+
+		result.Add(updatedDef)
+	}
+
+	return result, nil
+}
+
+func removeStatusSecrets(types astmodel.Types) (astmodel.Types, error) {
+	specVisitor := astmodel.TypeVisitorBuilder{
+		VisitObjectType: removeSecretProperties,
+	}.Build()
+
+	statusTypes, err := astmodel.FindStatusConnectedTypes(types)
+	if err != nil {
+		return nil, errors.Wrap(err, "couldn't find all status types")
+	}
+
+	result := make(astmodel.Types)
+
+	for _, def := range statusTypes {
+		updatedDef, err := specVisitor.VisitDefinition(def, nil)
+		if err != nil {
+			return nil, errors.Wrapf(err, "visiting type %q", def.Name())
+		}
+
+		result.Add(updatedDef)
+	}
+
+	return result, nil
+}
+
+func isTypeString(t astmodel.Type) bool {
+	primType, ok := astmodel.AsPrimitiveType(t)
+	return ok && primType == astmodel.StringType
+}
+
+func removeSecretProperties(_ *astmodel.TypeVisitor, it *astmodel.ObjectType, _ interface{}) (astmodel.Type, error) {
+	for _, prop := range it.Properties() {
+		if prop.IsSecret() {
+			// The expectation is that this is a string
+			propType := prop.PropertyType()
+			if !isTypeString(propType) {
+				return nil, errors.Errorf("expected property %q to be a string, but was: %T", prop.PropertyName(), propType)
+			}
+
+			it = it.WithoutProperty(prop.PropertyName())
+		}
+	}
+
+	return it, nil
+}
+
+func transformSecretProperties(_ *astmodel.TypeVisitor, it *astmodel.ObjectType, _ interface{}) (astmodel.Type, error) {
+	for _, prop := range it.Properties() {
+		if prop.IsSecret() {
+			// The expectation is that this is a string
+			propType := prop.PropertyType()
+			if !isTypeString(propType) {
+				return nil, errors.Errorf("expected property %q to be a string, but was: %T", prop.PropertyName(), propType)
+			}
```

Comment:
With `PropertyDefinition.HasType()`, this would become

```suggestion
			if !prop.HasType(astmodel.StringType) {
				return nil, errors.Errorf("expected property %q to be a string, but was: %T", prop.PropertyName(), prop.PropertyType())
			}
```

Reply:
See above